File: Leetcode/Algorithm/python/0000/00651-4 Keys Keyboard.py
class Solution(object):
    def maxA(self, n):
        self.d = {}
        return self.dfs(n)
        
    def dfs(self, n):
        if n == 1:
            return n
        if n in self.d:
            return self.d[n]
        
        ans = n
        for i in xrange(1, n - 3):
            ans = max(ans, self.dfs(i) * (n - i - 1))
        self.d[n] = ans
        return ans

File: Leetcode/Algorithm/cpp/0000/00651-4 Keys Keyboard.cc
const int SIZE = 55;

class Solution {
public:
    int maxA(int N) {
        vector<int> dp(SIZE, 0);
        
        for (int i = 1; i < SIZE; i++) {
            dp[i] = max(dp[i], dp[i - 1] + 1); // add A
            if (i - 3 >= 0) {
                int cnt = dp[i - 3];
                for (int j = 0; j + i < SIZE; j++) {
                    dp[i + j] = max(dp[i + j], cnt * (j + 2));
                }
            }
            // cout << dp[i] << endl;
        }
        
        return dp[N];
    }
};
